channel_provider.go

```
@@ -0,0 +1,444 @@
+			if err != nil {
+				logger.Error("failed to get cluster endpoints", slog.Any("error", err))
+				return
+			}
+
+			endpointsChan <- endpointsResp.Endpoints
+		}(conn)
+	}
+
+	go func() {
+		wg.Wait()
+		close(endpointsChan)
+	}()
+
+	// Stores the endpoints from the node with the largest view of the cluster
+	var maxTempEndpoints map[uint64]*protos.ServerEndpointList
+	for endpoints := range endpointsChan {
+		if maxTempEndpoints == nil || len(endpoints) > len(maxTempEndpoints) {
+			maxTempEndpoints = endpoints
+		}
+	}
+
+	return maxTempEndpoints
+}
+
+func (cp *ChannelProvider) checkAndSetNodeConns(newNodeEndpoints map[uint64]*protos.ServerEndpointList) {
+	wg := sync.WaitGroup{}
+
+	// Find which nodes have a different endpoint list and update their channel
+	for node, newEndpoints := range newNodeEndpoints {
+		wg.Add(1)
+
+		go func(node uint64, newEndpoints *protos.ServerEndpointList) {
+			defer wg.Done()
+
+			addNewConn := true
+
+			if currEndpoints, ok := cp.nodeConns[node]; ok {
+				if endpointListEqual(currEndpoints.Endpoints, newEndpoints) {
+					addNewConn = false
+				} else {
+					cp.logger.Debug("endpoints for node changed, recreating channel", slog.Uint64("node", node))
+
+					addNewConn = true
+
+					currEndpoints.Channel.Close()
+				}
+			}
+
+			if addNewConn {
+				// Either this is a new node or its endpoints have changed
+				err := cp.updateNodeConns(context.TODO(), node, newEndpoints)
+				if err != nil {
+					cp.logger.Error("failed to create new channel", slog.Uint64("node", node), slog.Any("error", err))
+				}
+			}
+		}(node, newEndpoints)
+	}
+
+	wg.Wait()
+}
+
+func (cp *ChannelProvider) removeDownNodes(newNodeEndpoints map[uint64]*protos.ServerEndpointList) {
+	cp.tendLock.Lock()
+
+	// The cluster state changed. Remove old channels.
+	for node, channelEndpoints := range cp.nodeConns {
+		if _, ok := newNodeEndpoints[node]; !ok {
+			channelEndpoints.Channel.Close()
```

The returned error is not checked.

---

which returned error? Are you talking about tend() not returning an error?

---

`channelEndpoints.Channel.Close()` returns an error.

---

sorry, I was using vscode to review your comments and there was a mismatch
between the line numbers and the comments